### PR TITLE
[mle] implement periodic parent search mechanism on FED/REED

### PR DESCRIPTION
**Pull request by abtink**
This commit implements "Periodic Parent Search" mechanism for FED/REED devices (FTD children). This enhances and builds upon the existing parent search feature, which is mainly intended for MTD children.

An FTD child receives and processes MLE Advertisements from neighboring routers. The child uses this information to track the one-way link quality to each router, which is later used to compare and select potential new parents.

Every "parent search check interval", the FTD child checks to see if it can select a better parent by evaluating all neighboring routers based on their link quality information. A router is considered a suitable parent candidate only if its average RSS exceeds the current parent's RSS by a margin (`PARENT_SEARCH_RSS_MARGIN`).

Once a candidate is selected, the FTD child sends unicast MLE Parent Requests to both the candidate and its current parent. This ensures updated connectivity information is obtained from both before making a decision. The same set of criteria used to compare candidates during initial attach are applied during parent switch.

If the attach attempt to the selected candidate fails (e.g., the router already has the maximum number of children it can support), the FTD child ensures not to select the same router again until a "reselect timeout" expires.

This commit also adds the `test-034-fed-parent-search.py test`, which validates the newly added FTD parent search behavior, including the mechanisms to attach to a selected router and the reselect timeout.

----

Related to [SPEC-1224](https://threadgroup.atlassian.net/browse/SPEC-1224).

----

_Copying updated documentation (from `config/parent_search.h`):_

The parent search mechanism depends on whether the device is an FTD child or an MTD child.

### FTD Child

- An FTD child receives and processes MLE Advertisements from neighboring routers. It uses this information to track the one-way link quality to each, which is later used to compare and select potential new parents.
- Every `OPENTHREAD_CONFIG_PARENT_SEARCH_CHECK_INTERVAL` seconds, an FTD child tries to select a better parent. The FTD child checks the list of neighboring routers and the tracked link quality information. A new parent is selected only if its average RSS exceeds the current parent's RSS by a margin specified by the configuration `OPENTHREAD_CONFIG_PARENT_SEARCH_RSS_MARGIN` configuration.
- If the attach attempt to the selected router fails (e.g., the router already has the maximum number of children it can support), the FTD child ensures that the same router cannot be selected again until a "reselect timeout" expires. This avoids repeated attempts to the same router. This timeout is specified by the configuration `OPENTHREAD_CONFIG_PARENT_SEARCH_RESELECT_TIMEOUT`.

### MTD Child

- Every `OPENTHREAD_CONFIG_PARENT_SEARCH_CHECK_INTERVAL` seconds, an MTD child checks its average RSS to its current parent. The child starts a parent search process only if the average RSS is below `OPENTHREAD_CONFIG_PARENT_SEARCH_RSS_THRESHOLD`.
- This ensures that an MTD child already attached to a parent with good link quality does not waste energy searching for better parents.
- The MTD child sends an MLE Parent Request to discover possible new parents. Because this process can be power-consuming (the child needs to stay in RX mode to collect parent responses), and to limit its impact on battery-powered devices, after a parent search is triggered on an MTD, the MTD child does not trigger another one before the specified backoff interval (`OPENTHREAD_CONFIG_PARENT_SEARCH_BACKOFF_INTERVAL`) expires.

### CODE REVIEW

File: src/core/config/parent_search.h [hunk 1]
```diff
@@ -48,16 +48,29 @@
...
- * better parent and will switch parent if a better one is found.
+ * When this feature is enabled, an end device/child (while staying attached) periodically searches for a possible
+ * better parent and switches parents if a better one is found.
  *
- * The child will periodically check the average RSS value for the current parent, and only if it is below a specific
- * threshold, a parent search is performed. The `OPENTHREAD_CONFIG_PARENT_SEARCH_CHECK_INTERVAL` specifies the
- * check interval (in seconds) and `OPENTHREAD_CONFIG_PARENT_SEARCH_RSS_THRESHOLD` gives the RSS threshold.
+ * Every `OPENTHREAD_CONFIG_PARENT_SEARCH_CHECK_INTERVAL` seconds, the child checks its average RSS to its current
+ * parent. Only if the average RSS is below a configured threshold (`OPENTHREAD_CONFIG_PARENT_SEARCH_RSS_THRESHOLD`)
+ * does the device start a parent search process.
```
Comment (jwhui) [inline src/core/config/parent_search.h:56]: Why not always try to switch parents if there is a neighboring router that has a much better signal strength?

Comment (abtink) [reply src/core/config/parent_search.h:56]: This rule applies to both MTD and FTD.

For MTD (SED), we used this to skip the entire search process and save battery life if the child is happy with its current parent (has a decent RSSI). For FTD, I agree we can consider relaxing this. 

The way I would suggest is to use different (default) threshold values. By setting `OPENTHREAD_CONFIG_PARENT_SEARCH_RSS_THRESHOLD` to a large value on FTD, we can effectively disable this check. This approach offers flexibility while reusing the same logic and code.

Thoughts?

Comment (jwhui) [reply src/core/config/parent_search.h:56]: Yes, having a different default values makes sense. I would suggest setting the FTD to disable this check.

Comment (abtink) [reply src/core/config/parent_search.h]: Updated the PR. Opted to fully remove the current "Parent RSS" check for FTD child.

File: src/core/config/parent_search.h [hunk 2]
```diff
@@ -46,18 +46,37 @@
...
  *
- * The child will periodically check the average RSS value for the current parent, and only if it is below a specific
- * threshold, a parent search is performed. The `OPENTHREAD_CONFIG_PARENT_SEARCH_CHECK_INTERVAL` specifies the
- * check interval (in seconds) and `OPENTHREAD_CONFIG_PARENT_SEARCH_RSS_THRESHOLD` gives the RSS threshold.
+ * The parent search mechanism depends on whether the device is an FTD child or an MTD child.
  *
- * Since the parent search process can be power consuming (child needs to stays in RX mode to collect parent response)
- * and to limit its impact on battery-powered devices, after a parent search is triggered, the child will not trigger
- * another one before a specified backoff interval specified by `OPENTHREAD_CONFIG_PARENT_SEARCH_BACKOFF_INTERVAL`.
+ * FTD Child
```
Comment (jwhui) [inline src/core/config/parent_search.h:56]: Should we consider enabling this by default on an FTD child?

Comment (abtink) [reply src/core/config/parent_search.h:56]: Yes, that makes sense. I have made the change in a new push and added a comment to encourage enabling this feature on MTD builds as well, taking into account the parameters that make sense for a battery-powered device based on its battery capacity and intended battery life.

**Review by jwhui — APPROVED** [review]

**Pull request merged**